src/dist/replication/lib/replica_stub.cpp

```diff
@@ -889,6 +889,87 @@ void replica_stub::on_query_replica_info(const query_replica_info_request &req,
     resp.err = ERR_OK;
 }
 
+void replica_stub::on_query_disk_info(const query_disk_info_request &req,
+                                      /*out*/ query_disk_info_response &resp)
+{
+    ddebug("got query disk info request from (%s)", req.node.to_string());
+    _fs_manager.compute_disk_density();
+
+    std::set<app_id> app_ids;
+    {
+        zauto_read_lock l(_replicas_lock);
+        for (auto it = _replicas.begin(); it != _replicas.end(); ++it) {
+            replica_ptr &r = it->second;
+            const app_info &a_info = *r->get_app_info();
+            if (app_ids.find(a_info.app_id) == app_ids.end()) {
+                app_ids.insert(a_info.app_id);
+            }
+        }
+    }
+
+    for (auto &n : _fs_manager._dir_nodes) {
+        disk_info d_info;
+        for (auto app_id : app_ids) {
+            const auto &iter = n->holding_replicas.find(app_id);
+            if (iter == n->holding_replicas.end()) {
+                dwarn("disk(disk_dir = %s) hasn't been allocated space for app(app_id = %d)",
```

Comment:
It's normal that a dir not contain an app, why logging as warning? Suppose a cluster with many 4-partion apps, there will be many WARN level log.